src/Common/ARCFileCache.cpp

```diff
@@ -0,0 +1,971 @@
+            }
+            case FileSegment::State::DOWNLOADING: {
+                ++stat.downloading_size;
+                break;
+            }
+            default:
+                break;
+        }
+    }
+    return stat;
+}
+
+void ARCFileCache::reduceSizeToDownloaded(
+    const Key & key, size_t offset, std::lock_guard<std::mutex> & cache_lock, std::lock_guard<std::mutex> & /* segment_lock */)
+{
+    /**
+     * In case file was partially downloaded and it's download cannot be continued
+     * because of no space left in cache, we need to be able to cut cell's size to downloaded_size.
+     */
+
+    auto * cell = getCell(key, offset, cache_lock);
+
+    if (!cell)
+        throw Exception(ErrorCodes::REMOTE_FS_OBJECT_CACHE_ERROR, "No cell found for key: {}, offset: {}", keyToStr(key), offset);
+
+    const auto & file_segment = cell->file_segment;
+
+    size_t downloaded_size = file_segment->downloaded_size;
+    if (downloaded_size == file_segment->range().size())
+        throw Exception(
+            ErrorCodes::REMOTE_FS_OBJECT_CACHE_ERROR,
+            "Nothing to reduce, file segment fully downloaded, key: {}, offset: {}",
+            keyToStr(key),
+            offset);
+
+    cell->file_segment = std::make_shared<FileSegment>(offset, downloaded_size, key, this, FileSegment::State::DOWNLOADED);
+}
+
+bool ARCFileCache::isLastFileSegmentHolder(
+    const Key & key, size_t offset, std::lock_guard<std::mutex> & cache_lock, std::lock_guard<std::mutex> & /* segment_lock */)
+{
+    auto * cell = getCell(key, offset, cache_lock);
+
+    if (!cell)
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "No cell found for key: {}, offset: {}", keyToStr(key), offset);
+
+    /// The caller of this method is last file segment holder if use count is 2 (the second pointer is cache itself)
+    return cell->file_segment.use_count() == 2;
+}
+
+bool ARCFileCache::canMoveCellToHighQueue(const FileSegmentCell & cell) const
+{
+    return cell.hit_count >= move_threshold;
+}
+
+bool ARCFileCache::tryMoveLowToHigh(const FileSegmentCell & cell, std::lock_guard<std::mutex> & cache_lock)
+{
+    int increase_queue_size = 0;
+    int increase_space_bytes = 0;
+
+    if (high_queue.getMaxSpaceBytes() < max_high_space_size)
+    {
+        increase_space_bytes = std::min(cell.size(), max_high_space_size - high_queue.getMaxSpaceBytes());
+        high_queue.incrementMaxSpaceBytes(increase_space_bytes);
+        low_queue.incrementMaxSpaceBytes(0 - increase_space_bytes);
+    }
+
+    if (high_queue.getQueueSize() > high_queue.getMaxQueueSize())
+    {
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "Large cache queue size. There must be a bug");
+    }
+
+    /// Determine the need to adjust the size of the high queue
+    if ((high_queue.getQueueSize() == high_queue.getMaxQueueSize()) && (high_queue.getMaxQueueSize() < max_high_elem_size))
+    {
+        increase_queue_size = 1;
+        high_queue.incrementMaxQueueSize(increase_queue_size);
```

Comment:
it seems that it should be possible to shrink it back (on elements removal from high queue), now looks like as it grows, it will never be lower.